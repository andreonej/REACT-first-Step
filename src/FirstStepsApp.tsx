import { useState } from 'react';
import { ItemCounter } from './shopping-cart/ItemCounter';
import styles from './FirstStepsApp.module.css';

// Interfaz mejorada para los productos
interface Product {
  id: string;
  productName: string;
  quantity: number;
  price: number;
  category: string;
}

// Datos de ejemplo mÃ¡s realistas
const initialProducts: Product[] = [
  { 
    id: '1', 
    productName: 'Nintendo Switch OLED', 
    quantity: 1, 
    price: 349.99,
    category: 'Consolas'
  },
  { 
    id: '2', 
    productName: 'Pro Controller', 
    quantity: 2, 
    price: 69.99,
    category: 'Accesorios'
  },
  { 
    id: '3', 
    productName: 'Super Smash Bros Ultimate', 
    quantity: 1, 
    price: 59.99,
    category: 'Juegos'
  },
];

export function FirstStepsApp() {
  const [products, setProducts] = useState<Product[]>(initialProducts);

  const handleQuantityChange = (productId: string, newQuantity: number) => {
    setProducts(prevProducts => 
      prevProducts.map(product => 
        product.id === productId 
          ? { ...product, quantity: newQuantity }
          : product
      )
    );
  };

  const totalItems = products.reduce((sum, product) => sum + product.quantity, 0);
  const totalPrice = products.reduce((sum, product) => sum + (product.quantity * product.price), 0);

  const clearCart = () => {
    setProducts([]);
  };

  return (
    <div className={styles.container}>
      <header className={styles.header}>
        <h1 className={styles.title}>ðŸ›’ Carrito de Compras</h1>
        <div className={styles.summary}>
          <span className={styles.itemCount}>{totalItems} artÃ­culos</span>
          <span className={styles.totalPrice}>${totalPrice.toFixed(2)}</span>
        </div>
      </header>

      <main className={styles.main}>
        {products.length > 0 ? (
          <>
            <div className={styles.productList}>
              {products.map((product) => (
                <ItemCounter 
                  key={product.id}
                  name={product.productName}
                  quantity={product.quantity}
                  maxQuantity={10}
                  onQuantityChange={(newQuantity) => 
                    handleQuantityChange(product.id, newQuantity)
                  }
                />
              ))}
            </div>
            
            <footer className={styles.footer}>
              <button 
                className={styles.clearButton}
                onClick={clearCart}
              >
                Vaciar Carrito
              </button>
              <button className={styles.checkoutButton}>
                Proceder al Pago - ${totalPrice.toFixed(2)}
              </button>
            </footer>
          </>
        ) : (
          <div className={styles.emptyCart}>
            <div className={styles.emptyIcon}>ðŸ›’</div>
            <h2>Tu carrito estÃ¡ vacÃ­o</h2>
            <p>Agrega algunos productos para comenzar</p>
          </div>
        )}
      </main>
    </div>
  );
}
