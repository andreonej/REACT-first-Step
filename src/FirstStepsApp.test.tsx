/* eslint-disable @typescript-eslint/no-unused-vars */
import { render, screen } from '@testing-library/react';
import { afterEach, describe, expect, test, vi } from 'vitest';
import { FirstStepsApp } from './FirstStepsApp';

const mockItemCounter = vi.fn((_props: unknown) => {
  return <div data-testid="ItemCounter" />;
});

vi.mock('./shopping-cart/ItemCounter', () => ({
  ItemCounter: (props: unknown) => mockItemCounter(props),
}));

describe('FirstStepsApp', () => {
  //Antes de cada prueba, limpiamos los mocks para evitar que los resultados de una prueba afecten a otra.
  afterEach(() => {
    vi.clearAllMocks(); // para limpiar los mocks despuÃ©s de cada prueba y evitar conteos incorrectos.
  });

  test('should render the correct number of ItemCounter components', () => {
    render(<FirstStepsApp />);
    const itemCounters = screen.getAllByTestId('ItemCounter');
    expect(itemCounters.length).toBe(3);
  });

  test('should render ItemCounter with correct props', () => {
    render(<FirstStepsApp />);

    // Verificamos que se haya llamado 3 veces (una por cada producto)
    expect(mockItemCounter).toHaveBeenCalledTimes(3);

    // Verificamos la primera llamada (Nintendo Switch OLED)
    expect(mockItemCounter).toHaveBeenNthCalledWith(1, expect.objectContaining({
      name: 'Nintendo Switch OLED',
      quantity: 1,
      maxQuantity: 10,
      onQuantityChange: expect.any(Function)
    }));

    // Verificamos la segunda llamada (Pro Controller)
    expect(mockItemCounter).toHaveBeenNthCalledWith(2, expect.objectContaining({
      name: 'Pro Controller',
      quantity: 2,
      maxQuantity: 10,
      onQuantityChange: expect.any(Function)
    }));

    // Verificamos la tercera llamada (Super Smash Bros Ultimate)
    expect(mockItemCounter).toHaveBeenNthCalledWith(3, expect.objectContaining({
      name: 'Super Smash Bros Ultimate',
      quantity: 1,
      maxQuantity: 10,
      onQuantityChange: expect.any(Function)
    }));
  });
});
