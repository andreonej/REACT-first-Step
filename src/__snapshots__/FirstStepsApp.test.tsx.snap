// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepsApp > should match snapshot 1`] = `
<div>
  <div
    class="_container_fff967"
  >
    <header
      class="_header_fff967"
    >
      <h1
        class="_title_fff967"
      >
        ðŸ›’ Carrito de Compras
      </h1>
      <div
        class="_summary_fff967"
      >
        <span
          class="_itemCount_fff967"
        >
          4
           artÃ­culos
        </span>
        <span
          class="_totalPrice_fff967"
        >
          $
          549.96
        </span>
      </div>
    </header>
    <main
      class="_main_fff967"
    >
      <div
        class="_productList_fff967"
      >
        <article
          class="_itemCounter_8b32a2"
        >
          <div
            class="_productInfo_8b32a2"
          >
            <h3
              class="_productName_8b32a2 _lowStock_8b32a2"
            >
              Nintendo Switch OLED
            </h3>
            <span
              class="_stockWarning_8b32a2"
            >
              Stock mÃ­nimo
            </span>
          </div>
          <div
            class="_quantityControls_8b32a2"
          >
            <button
              aria-label="Disminuir cantidad de Nintendo Switch OLED"
              class="_button_8b32a2 _decrementButton_8b32a2"
              disabled=""
            >
              âˆ’
            </button>
            <span
              aria-label="Cantidad: 1"
              class="_quantity_8b32a2"
            >
              1
            </span>
            <button
              aria-label="Aumentar cantidad de Nintendo Switch OLED"
              class="_button_8b32a2 _incrementButton_8b32a2"
            >
              +
            </button>
          </div>
        </article>
        <article
          class="_itemCounter_8b32a2"
        >
          <div
            class="_productInfo_8b32a2"
          >
            <h3
              class="_productName_8b32a2 "
            >
              Pro Controller
            </h3>
          </div>
          <div
            class="_quantityControls_8b32a2"
          >
            <button
              aria-label="Disminuir cantidad de Pro Controller"
              class="_button_8b32a2 _decrementButton_8b32a2"
            >
              âˆ’
            </button>
            <span
              aria-label="Cantidad: 2"
              class="_quantity_8b32a2"
            >
              2
            </span>
            <button
              aria-label="Aumentar cantidad de Pro Controller"
              class="_button_8b32a2 _incrementButton_8b32a2"
            >
              +
            </button>
          </div>
        </article>
        <article
          class="_itemCounter_8b32a2"
        >
          <div
            class="_productInfo_8b32a2"
          >
            <h3
              class="_productName_8b32a2 _lowStock_8b32a2"
            >
              Super Smash Bros Ultimate
            </h3>
            <span
              class="_stockWarning_8b32a2"
            >
              Stock mÃ­nimo
            </span>
          </div>
          <div
            class="_quantityControls_8b32a2"
          >
            <button
              aria-label="Disminuir cantidad de Super Smash Bros Ultimate"
              class="_button_8b32a2 _decrementButton_8b32a2"
              disabled=""
            >
              âˆ’
            </button>
            <span
              aria-label="Cantidad: 1"
              class="_quantity_8b32a2"
            >
              1
            </span>
            <button
              aria-label="Aumentar cantidad de Super Smash Bros Ultimate"
              class="_button_8b32a2 _incrementButton_8b32a2"
            >
              +
            </button>
          </div>
        </article>
      </div>
      <footer
        class="_footer_fff967"
      >
        <button
          class="_clearButton_fff967"
        >
          Vaciar Carrito
        </button>
        <button
          class="_checkoutButton_fff967"
        >
          Proceder al Pago - $
          549.96
        </button>
      </footer>
    </main>
  </div>
</div>
`;
